File: ts/src/schema.ts
/**
 * The `envlint.toml` schema model and the validation engine.
 */

import { parse as parseToml } from "smol-toml";

import { Report, errorIssue, warningIssue } from "./report.ts";
import {
  type VarType,
  VAR_TYPES,
  DEFAULT_VAR_TYPE,
  coerce,
  asNumber,
} from "./value.ts";

/**
 * The declared constraints for a single variable. This is the programmatic
 * shape; the TOML loader produces the same object after validation.
 */
export interface VarSpec {
  /** Variable type. Defaults to `"string"` when omitted. */
  readonly type?: VarType;
  /** When true, the variable must be present and non-empty. */
  readonly required?: boolean;
  /** Value used when the variable is absent from the environment. */
  readonly default?: string;
  /** Regex (matched on the raw text) the value must satisfy. */
  readonly pattern?: string;
  /** Allowed values for an `enum`-typed variable. */
  readonly values?: readonly string[];
  /** Inclusive lower bound for numeric/duration values. */
  readonly min?: number;
  /** Inclusive upper bound for numeric/duration values. */
  readonly max?: number;
  /** When true, the resolved value is masked as `******` in all output. */
  readonly secret?: boolean;
  /** Human-readable description (informational only). */
  readonly description?: string;
}

/** A schema definition: a set of named variable specs plus a strict flag. */
export interface SchemaDef {
  readonly vars: Record<string, VarSpec>;
  /**
   * When true, variables present in the environment but absent from the schema
   * are reported as errors instead of warnings.
   */
  readonly strict?: boolean;
}

/**
 * An error raised while *loading* a schema (as opposed to validating an
 * environment against it). Schema-authoring mistakes fail fast here so the
 * error points at the schema, not the user's environment.
 */
export class SchemaError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SchemaError";
  }
}

/**
 * A parsed, internally-consistent schema. Construct one from TOML with
 * {@link Schema.fromTomlStr} or from a TS object with {@link Schema.fromDef}.
 */
export class Schema {
  /** Variable specs, keyed by name. */
  readonly vars: Map<string, VarSpec>;
  /** Whether undeclared variables are errors (true) or warnings (false). */
  strict: boolean;
  /** Compiled patterns, cached per variable (validated at construction). */
  private readonly patterns = new Map<string, RegExp>();

  private constructor(vars: Map<string, VarSpec>, strict: boolean) {
    this.vars = vars;
    this.strict = strict;
    for (const [name, spec] of vars) {
      if (spec.pattern !== undefined) {
        this.patterns.set(name, compilePattern(name, spec.pattern));
      }
    }
  }

  /** Parse a schema from TOML source, validating it for internal consistency. */
  static fromTomlStr(src: string): Schema {
    let parsed: unknown;
    try {
      parsed = parseToml(src);
    } catch (e) {
      throw new SchemaError(`invalid schema: ${(e as Error).message}`);
    }
    return Schema.fromDef(coerceTomlToDef(parsed));
  }

  /**
   * Build a schema from a programmatic definition, validating it for internal
   * consistency (patterns compile, enums declare values, types are known).
   */
  static fromDef(def: SchemaDef): Schema {
    const vars = new Map<string, VarSpec>();
    // Iterate in sorted name order so reports are deterministic regardless of
    // how the definition was authored (mirrors the Rust BTreeMap).
    for (const name of Object.keys(def.vars).sort()) {
      const spec = def.vars[name];
      if (spec === undefined) continue;
      validateSpec(name, spec);
      vars.set(name, spec);
    }
    return new Schema(vars, def.strict ?? false);
  }

  /**
   * Validate a set of variables (parsed from a `.env` file or read from the
   * process environment) against this schema, returning a {@link Report}.
   */
  validate(env: Record<string, string>): Report {
    const report = new Report();

    for (const [name, spec] of this.vars) {
      if (spec.secret) report.markSecret(name);

      // Resolve the raw value: an explicit value wins over the default.
      const explicit = Object.prototype.hasOwnProperty.call(env, name)
        ? env[name]
        : undefined;
      const raw = explicit ?? spec.default;

      if (raw === undefined) {
        if (spec.required) {
          report.add(errorIssue(name, "required variable is not set"));
        }
        continue;
      }

      // An explicitly-set-but-empty value for a required var is an error.
      if (spec.required && raw.length === 0) {
        report.add(errorIssue(name, "required variable is empty"));
        continue;
      }

      this.checkVar(name, spec, raw, report);
    }

    // Report variables present in the environment but not in the schema, in
    // sorted order for deterministic output.
    for (const name of Object.keys(env).sort()) {
      if (!this.vars.has(name)) {
        report.add(
          this.strict
            ? errorIssue(name, "variable is not declared in the schema")
            : warningIssue(name, "variable is not declared in the schema"),
        );
      }
    }

    return report;
  }

  private checkVar(name: string, spec: VarSpec, raw: string, report: Report): void {
    const ty = spec.type ?? DEFAULT_VAR_TYPE;
    const values = spec.values ?? [];

    // Enum membership is checked on the raw string.
    if (ty === "enum" && !values.includes(raw)) {
      report.add(
        errorIssue(name, `must be one of ${formatList(values)}, got ${quote(raw)}`),
      );
      return;
    }

    const result = coerce(raw, ty);
    if (!result.ok) {
      report.add(errorIssue(name, result.error));
      return;
    }
    const value = result.value;

    // `pattern` applies to the raw textual form.
    const re = this.patterns.get(name);
    if (re !== undefined && !re.test(raw)) {
      report.add(errorIssue(name, `does not match pattern /${spec.pattern}/`));
      return;
    }

    // `min`/`max` apply to numeric/duration values.
    const n = asNumber(value);
    if (n !== undefined) {
      if (spec.min !== undefined && n < spec.min) {
        report.add(errorIssue(name, `${formatNumber(n)} is below min ${formatNumber(spec.min)}`));
        return;
      }
      if (spec.max !== undefined && n > spec.max) {
        report.add(errorIssue(name, `${formatNumber(n)} is above max ${formatNumber(spec.max)}`));
        return;
      }
    }

    report.resolve(name, value);
  }
}

/** Compile a pattern, raising a {@link SchemaError} attributed to its var. */
function compilePattern(name: string, pattern: string): RegExp {
  try {
    return new RegExp(pattern);
  } catch (e) {
    throw new SchemaError(`${name}: invalid pattern: ${(e as Error).message}`);
  }
}

/** Validate a single spec for internal consistency. */
function validateSpec(name: string, spec: VarSpec): void {
  const ty = spec.type ?? DEFAULT_VAR_TYPE;
  if (!VAR_TYPES.includes(ty)) {
    throw new SchemaError(`${name}: unknown type ${quote(ty)}`);
  }
  if (spec.pattern !== undefined) {
    compilePattern(name, spec.pattern);
  }
  if (ty === "enum" && (spec.values === undefined || spec.values.length === 0)) {
    throw new SchemaError(`${name}: enum type requires a non-empty \`values\` list`);
  }
}

/**
 * Convert the loosely-typed result of TOML parsing into a {@link SchemaDef},
 * rejecting structurally-invalid documents with a {@link SchemaError}.
 */
function coerceTomlToDef(parsed: unknown): SchemaDef {
  if (parsed === null || typeof parsed !== "object") {
    throw new SchemaError("invalid schema: expected a TOML table");
  }
  const root = parsed as Record<string, unknown>;

  const strict = root["strict"];
  if (strict !== undefined && typeof strict !== "boolean") {
    throw new SchemaError("invalid schema: `strict` must be a boolean");
  }

  const rawVars = root["vars"] ?? {};
  if (rawVars === null || typeof rawVars !== "object") {
    throw new SchemaError("invalid schema: `vars` must be a table");
  }

  const vars: Record<string, VarSpec> = {};
  for (const [name, rawSpec] of Object.entries(rawVars as Record<string, unknown>)) {
    vars[name] = coerceTomlToSpec(name, rawSpec);
  }

  return { vars, strict: strict as boolean | undefined };
}

function coerceTomlToSpec(name: string, raw: unknown): VarSpec {
  if (raw === null || typeof raw !== "object") {
    throw new SchemaError(`invalid schema: [vars.${name}] must be a table`);
  }
  const t = raw as Record<string, unknown>;

  const spec: {
    type?: VarType;
    required?: boolean;
    default?: string;
    pattern?: string;
    values?: string[];
    min?: number;
    max?: number;
    secret?: boolean;
    description?: string;
  } = {};

  if (t["type"] !== undefined) {
    if (typeof t["type"] !== "string") {
      throw new SchemaError(`${name}: \`type\` must be a string`);
    }
    spec.type = t["type"] as VarType;
  }
  if (t["required"] !== undefined) {
    if (typeof t["required"] !== "boolean") {
      throw new SchemaError(`${name}: \`required\` must be a boolean`);
    }
    spec.required = t["required"];
  }
  if (t["default"] !== undefined) {
    spec.default = stringifyScalar(name, "default", t["default"]);
  }
  if (t["pattern"] !== undefined) {
    if (typeof t["pattern"] !== "string") {
      throw new SchemaError(`${name}: \`pattern\` must be a string`);
    }
    spec.pattern = t["pattern"];
  }
  if (t["values"] !== undefined) {
    if (!Array.isArray(t["values"]) || t["values"].some((v) => typeof v !== "string")) {
      throw new SchemaError(`${name}: \`values\` must be an array of strings`);
    }
    spec.values = t["values"] as string[];
  }
  if (t["min"] !== undefined) {
    if (typeof t["min"] !== "number") {
      throw new SchemaError(`${name}: \`min\` must be a number`);
    }
    spec.min = t["min"];
  }
  if (t["max"] !== undefined) {
    if (typeof t["max"] !== "number") {
      throw new SchemaError(`${name}: \`max\` must be a number`);
    }
    spec.max = t["max"];
  }
  if (t["secret"] !== undefined) {
    if (typeof t["secret"] !== "boolean") {
      throw new SchemaError(`${name}: \`secret\` must be a boolean`);
    }
    spec.secret = t["secret"];
  }
  if (t["description"] !== undefined) {
    if (typeof t["description"] !== "string") {
      throw new SchemaError(`${name}: \`description\` must be a string`);
    }
    spec.description = t["description"];
  }

  return spec;
}

/**
 * The schema's `default` is a string in the TOML, but authors occasionally
 * write `default = 8080`. Accept scalar numbers/booleans and stringify them so
 * the value flows through coercion exactly as a quoted string would.
 */
function stringifyScalar(name: string, key: string, value: unknown): string {
  switch (typeof value) {
    case "string":
      return value;
    case "number":
      return formatNumber(value);
    case "boolean":
      return String(value);
    case "bigint":
      return value.toString();
    default:
      throw new SchemaError(`${name}: \`${key}\` must be a string, number, or boolean`);
  }
}

/** Quote a string the way Rust's `{:?}` does, for parity in messages. */
function quote(s: string): string {
  return JSON.stringify(s);
}

/** Render a string list the way Rust's `{:?}` renders a `Vec<String>`. */
function formatList(values: readonly string[]): string {
  return `[${values.map(quote).join(", ")}]`;
}

/** Render a number the way Rust's `{}` does (no trailing `.0` on integers). */
function formatNumber(n: number): string {
  return Number.isInteger(n) ? String(n) : String(n);
}

File: rust/src/schema.rs
//! The `envlint.toml` schema model and the validation engine.

use std::collections::BTreeMap;

use regex::Regex;
use serde::Deserialize;

use crate::report::{Issue, Report};
use crate::value::{coerce, VarType};

/// A parsed schema: a set of named variable specs.
#[derive(Debug, Deserialize)]
pub struct Schema {
    #[serde(default)]
    pub vars: BTreeMap<String, VarSpec>,
    /// When true, variables present in the environment but absent from the
    /// schema are reported as errors instead of warnings.
    #[serde(default)]
    pub strict: bool,
}

/// The declared constraints for a single variable.
#[derive(Debug, Deserialize)]
pub struct VarSpec {
    #[serde(rename = "type", default)]
    pub ty: VarType,
    #[serde(default)]
    pub required: bool,
    pub default: Option<String>,
    pub pattern: Option<String>,
    #[serde(default)]
    pub values: Vec<String>,
    pub min: Option<f64>,
    pub max: Option<f64>,
    #[serde(default)]
    pub secret: bool,
    pub description: Option<String>,
}

/// Errors that can occur while loading a schema (as opposed to validating
/// against it).
#[derive(Debug)]
pub enum SchemaError {
    Toml(toml::de::Error),
    /// A `pattern` failed to compile, for the named variable.
    BadPattern {
        var: String,
        message: String,
    },
    /// An `enum`-typed var declared no `values`.
    EmptyEnum {
        var: String,
    },
}

impl std::fmt::Display for SchemaError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SchemaError::Toml(e) => write!(f, "invalid schema: {e}"),
            SchemaError::BadPattern { var, message } => {
                write!(f, "{var}: invalid pattern: {message}")
            }
            SchemaError::EmptyEnum { var } => {
                write!(f, "{var}: enum type requires a non-empty `values` list")
            }
        }
    }
}

impl std::error::Error for SchemaError {}

impl Schema {
    /// Parse a schema from TOML source, validating it for internal consistency.
    pub fn from_toml_str(src: &str) -> Result<Schema, SchemaError> {
        let schema: Schema = toml::from_str(src).map_err(SchemaError::Toml)?;
        // Fail fast on schema-authoring mistakes so the error points at the
        // schema, not the user's environment.
        for (name, spec) in &schema.vars {
            if let Some(pat) = &spec.pattern {
                Regex::new(pat).map_err(|e| SchemaError::BadPattern {
                    var: name.clone(),
                    message: e.to_string(),
                })?;
            }
            if spec.ty == VarType::Enum && spec.values.is_empty() {
                return Err(SchemaError::EmptyEnum { var: name.clone() });
            }
        }
        Ok(schema)
    }

    /// Validate a set of variables (e.g. parsed from a `.env` file or read from
    /// the process environment) against this schema.
    pub fn validate(&self, env: &BTreeMap<String, String>) -> Report {
        let mut report = Report::default();

        for (name, spec) in &self.vars {
            if spec.secret {
                report.secrets.insert(name.clone(), ());
            }
            // Resolve the raw value: explicit value wins over default.
            let raw = env.get(name).cloned().or_else(|| spec.default.clone());

            let Some(raw) = raw else {
                if spec.required {
                    report
                        .issues
                        .push(Issue::error(name, "required variable is not set"));
                }
                continue;
            };

            // An explicitly-set-but-empty value for a required var is an error.
            if spec.required && raw.is_empty() {
                report
                    .issues
                    .push(Issue::error(name, "required variable is empty"));
                continue;
            }

            self.check_var(name, spec, &raw, &mut report);
        }

        // Report variables present in the environment but not in the schema.
        for name in env.keys() {
            if !self.vars.contains_key(name) {
                let issue = if self.strict {
                    Issue::error(name, "variable is not declared in the schema")
                } else {
                    Issue::warning(name, "variable is not declared in the schema")
                };
                report.issues.push(issue);
            }
        }

        report
    }

    fn check_var(&self, name: &str, spec: &VarSpec, raw: &str, report: &mut Report) {
        // Enum membership is checked on the raw string.
        if spec.ty == VarType::Enum && !spec.values.iter().any(|v| v == raw) {
            report.issues.push(Issue::error(
                name,
                format!("must be one of {:?}, got {:?}", spec.values, raw),
            ));
            return;
        }

        let value = match coerce(raw, spec.ty) {
            Ok(v) => v,
            Err(msg) => {
                report.issues.push(Issue::error(name, msg));
                return;
            }
        };

        // `pattern` applies to the raw textual form.
        if let Some(pat) = &spec.pattern {
            // Safe to unwrap: patterns are compiled in `from_toml_str`.
            let re = Regex::new(pat).expect("pattern validated at load time");
            if !re.is_match(raw) {
                report.issues.push(Issue::error(
                    name,
                    format!("does not match pattern /{pat}/"),
                ));
                return;
            }
        }

        // `min`/`max` apply to numeric/duration values.
        if let Some(n) = value.as_number() {
            if let Some(min) = spec.min {
                if n < min {
                    report
                        .issues
                        .push(Issue::error(name, format!("{n} is below min {min}")));
                    return;
                }
            }
            if let Some(max) = spec.max {
                if n > max {
                    report
                        .issues
                        .push(Issue::error(name, format!("{n} is above max {max}")));
                    return;
                }
            }
        }

        report.resolved.insert(name.to_string(), value);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn env(pairs: &[(&str, &str)]) -> BTreeMap<String, String> {
        pairs
            .iter()
            .map(|(k, v)| (k.to_string(), v.to_string()))
            .collect()
    }

    const SCHEMA: &str = r#"
        [vars.PORT]
        type = "port"
        default = "8080"

        [vars.LOG_LEVEL]
        type = "enum"
        values = ["debug", "info", "warn", "error"]
        default = "info"

        [vars.DATABASE_URL]
        type = "url"
        required = true

        [vars.API_KEY]
        type = "string"
        required = true
        secret = true
        pattern = "^sk-[A-Za-z0-9]{8,}$"

        [vars.WORKERS]
        type = "int"
        min = 1
        max = 64
        default = "4"
    "#;

    #[test]
    fn happy_path_uses_defaults() {
        let schema = Schema::from_toml_str(SCHEMA).unwrap();
        let report = schema.validate(&env(&[
            ("DATABASE_URL", "postgres://db:5432/app"),
            ("API_KEY", "sk-abcdef12"),
        ]));
        assert!(!report.has_errors(), "{:?}", report.issues);
        assert_eq!(report.resolved.len(), 5); // includes 3 defaults
    }

    #[test]
    fn flags_missing_required_and_bad_values() {
        let schema = Schema::from_toml_str(SCHEMA).unwrap();
        let report = schema.validate(&env(&[
            ("LOG_LEVEL", "verbose"),
            ("API_KEY", "nope"),
            ("WORKERS", "999"),
        ]));
        assert!(report.has_errors());
        let vars: Vec<_> = report.errors().map(|i| i.var.as_str()).collect();
        assert!(vars.contains(&"DATABASE_URL")); // required, missing
        assert!(vars.contains(&"LOG_LEVEL")); // not in enum
        assert!(vars.contains(&"API_KEY")); // pattern mismatch
        assert!(vars.contains(&"WORKERS")); // above max
    }

    #[test]
    fn undeclared_var_is_warning_then_error_in_strict() {
        let mut schema = Schema::from_toml_str(SCHEMA).unwrap();
        let e = env(&[
            ("DATABASE_URL", "https://x.y"),
            ("API_KEY", "sk-abcdef12"),
            ("MYSTERY", "1"),
        ]);
        let report = schema.validate(&e);
        assert!(!report.has_errors());
        assert_eq!(report.warnings().count(), 1);

        schema.strict = true;
        let report = schema.validate(&e);
        assert!(report.has_errors());
    }

    #[test]
    fn rejects_bad_schema() {
        assert!(Schema::from_toml_str("[vars.X]\ntype=\"string\"\npattern=\"(\"").is_err());
        assert!(Schema::from_toml_str("[vars.X]\ntype=\"enum\"").is_err());
    }
}
